<a href="https://colab.research.google.com/github/stephe-hu/datasci_2_manipulation/blob/main/hha507assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import re

# Loading file

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/stephe-hu/datasci_2_manipulation/main/data/healthcare_data_cleaning.csv')
df

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

# 1. Data Cleaning with Pandas

## Identify and handle missing values in the dataset

In [3]:
# Identifying and replace 'missing' values in columns NaN
df = df.replace('missing', np.nan)
# Drop rows with any NaN values
df.dropna(inplace=True)
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

## Remove any duplicate rows and columns, if they exist

In [4]:
# Identify and remove duplicate rows
df.drop_duplicates(inplace=True)
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       Average BP Height (in cm) Weight (in kg)     Payment Due ($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

## Clean column names

In [5]:
# Remove leading and trailing white space, convert to lowercases, and replace spaces with underscores for columns
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
df

patient_age  gender city_of_residence state_of_residence has_insurance  \
0               45   Other        West Brian           Kentucky           Yes   
4               48  Female         Birdmouth            Montana            No   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female  West Briannabury            Montana            No   
11              40  Female        Jonesmouth       Rhode Island            No   
...            ...     ...               ...                ...           ...   
104992          44  Female    Port Tracistad             Alaska           Yes   
104993          42  Female     West Annaport            Florida           Yes   
104994          53  Female     Michelleburgh       South Dakota            No   
104995          74   Other      West Anthony           Oklahoma           Yes   
104997          51  Female  Lake Lisachester           Nebraska           Yes   

       visited_last_month payment_method preferred_doctor disease_diagnosed  \
0                     Yes      Insurance     Dr. Williams               Flu   
4                      No      Insurance        Dr. Brown              None   
7                     Yes      Insurance     Dr. Williams               Flu   
8                     Yes           Card        Dr. Jones               Flu   
11                     No           Card        Dr. Jones          Covid-19   
...                   ...            ...              ...               ...   
104992                 No      Insurance        Dr. Brown              Cold   
104993                 No           Cash      Dr. Johnson          Covid-19   
104994                 No           Card     Dr. Williams               Flu   
104995                 No           Card     Dr. Williams          Covid-19   
104997                Yes      Insurance     Dr. Williams           Allergy   

       medication_prescribed type_of_appointment average_heart_rate  \
0                      Med_E             General                 66   
4                      Med_C          Specialist                 93   
7                      Med_E             General                 99   
8                      Med_C          Specialist                 67   
11                     Med_E          Specialist                 65   
...                      ...                 ...                ...   
104992                 Med_A           Follow-Up                 76   
104993                 Med_D          Specialist                 80   
104994                 Med_B           Emergency                 68   
104995                 Med_E          Specialist                 65   
104997                 Med_D           Follow-Up                 89   

       average_bp height_(in_cm) weight_(in_kg)     payment_due_($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       last_visit_(days_ago) visit_duration_(mins) number_of_tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193 

# 2. Data Transformation

## Create new columns based on existing ones

In [6]:
# Convert 'patient_age' to int and then create new column 'age_group'
df['patient_age'] = df['patient_age'].astype(int)
age_bins = [0, 30, 50, np.inf]
age_labels = ['Young', 'Middle-aged', 'Senior']
df['age_group'] = pd.cut(df['patient_age'], bins=age_bins, labels=age_labels, right=False)
df

patient_age  gender city_of_residence state_of_residence  \
0                45   Other        West Brian           Kentucky   
4                48  Female         Birdmouth            Montana   
7                52    Male      Vanessaburgh        Connecticut   
8                20  Female  West Briannabury            Montana   
11               40  Female        Jonesmouth       Rhode Island   
...             ...     ...               ...                ...   
104992           44  Female    Port Tracistad             Alaska   
104993           42  Female     West Annaport            Florida   
104994           53  Female     Michelleburgh       South Dakota   
104995           74   Other      West Anthony           Oklahoma   
104997           51  Female  Lake Lisachester           Nebraska   

       has_insurance visited_last_month payment_method preferred_doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
7                Yes                Yes      Insurance     Dr. Williams   
8                 No                Yes           Card        Dr. Jones   
11                No                 No           Card        Dr. Jones   
...              ...                ...            ...              ...   
104992           Yes                 No      Insurance        Dr. Brown   
104993           Yes                 No           Cash      Dr. Johnson   
104994            No                 No           Card     Dr. Williams   
104995           Yes                 No           Card     Dr. Williams   
104997           Yes                Yes      Insurance     Dr. Williams   

       disease_diagnosed medication_prescribed  ... average_heart_rate  \
0                    Flu                 Med_E  ...                 66   
4                   None                 Med_C  ...                 93   
7                    Flu                 Med_E  ...                 99   
8                    Flu                 Med_C  ...                 67   
11              Covid-19                 Med_E  ...                 65   
...                  ...                   ...  ...                ...   
104992              Cold                 Med_A  ...                 76   
104993          Covid-19                 Med_D  ...                 80   
104994               Flu                 Med_B  ...                 68   
104995          Covid-19                 Med_E  ...                 65   
104997           Allergy                 Med_D  ...                 89   

       average_bp height_(in_cm) weight_(in_kg)     payment_due_($)  \
0             111            183             70   65.45721578126224   
4              89            186             87   264.2147372473319   
7             129            170             47  387.08504353508863   
8             108            173             86   172.5483427968999   
11            106            166             82   299.1426761400895   
...           ...            ...            ...                 ...   
104992         93            178             66  462.56567087397013   
104993        129            171             78  131.35632576282202   
104994        109            172             75   488.0316041186948   
104995        112            169             61  309.16444606680665   
104997        118            179             71   320.8304180033907   

       last_visit_(days_ago) visit_duration_(mins) number_of_tests  \
0                        193                    92               5   
4                        188                   113               3   
7                        179                   108               2   
8                        174                    91               2   
11                       193                    92               1   
...                      ...                   ...             ...   
104992                   189                    55        

In [7]:
# Convert 'average_bp' to int and then create new column 'blood_pressure_status'
df['average_bp'] = df['average_bp'].astype(int)
threshold_high = 140
df['blood_pressure_status'] = df['average_bp'].apply(lambda x: 'High' if x >= threshold_high else 'Normal')
df

patient_age  gender city_of_residence state_of_residence  \
0                45   Other        West Brian           Kentucky   
4                48  Female         Birdmouth            Montana   
7                52    Male      Vanessaburgh        Connecticut   
8                20  Female  West Briannabury            Montana   
11               40  Female        Jonesmouth       Rhode Island   
...             ...     ...               ...                ...   
104992           44  Female    Port Tracistad             Alaska   
104993           42  Female     West Annaport            Florida   
104994           53  Female     Michelleburgh       South Dakota   
104995           74   Other      West Anthony           Oklahoma   
104997           51  Female  Lake Lisachester           Nebraska   

       has_insurance visited_last_month payment_method preferred_doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
7                Yes                Yes      Insurance     Dr. Williams   
8                 No                Yes           Card        Dr. Jones   
11                No                 No           Card        Dr. Jones   
...              ...                ...            ...              ...   
104992           Yes                 No      Insurance        Dr. Brown   
104993           Yes                 No           Cash      Dr. Johnson   
104994            No                 No           Card     Dr. Williams   
104995           Yes                 No           Card     Dr. Williams   
104997           Yes                Yes      Insurance     Dr. Williams   

       disease_diagnosed medication_prescribed  ... average_bp height_(in_cm)  \
0                    Flu                 Med_E  ...        111            183   
4                   None                 Med_C  ...         89            186   
7                    Flu                 Med_E  ...        129            170   
8                    Flu                 Med_C  ...        108            173   
11              Covid-19                 Med_E  ...        106            166   
...                  ...                   ...  ...        ...            ...   
104992              Cold                 Med_A  ...         93            178   
104993          Covid-19                 Med_D  ...        129            171   
104994               Flu                 Med_B  ...        109            172   
104995          Covid-19                 Med_E  ...        112            169   
104997           Allergy                 Med_D  ...        118            179   

        weight_(in_kg)     payment_due_($) last_visit_(days_ago)  \
0                   70   65.45721578126224                   193   
4                   87   264.2147372473319                   188   
7                   47  387.08504353508863                   179   
8                   86   172.5483427968999                   174   
11                  82   299.1426761400895                   193   
...                ...                 ...                   ...   
104992              66  462.56567087397013                   189   
104993              78  131.35632576282202                   185   
104994              75   488.0316041186948                   167   
104995              61  309.16444606680665                   183   
104997              71   320.8304180033907                   172   

       visit_duration_(mins) number_of_tests prescription_cost_($)  \
0                         92               5     15.71751735532889   
4                        113               3     77.48311304645532   
7                        108               2     76.96592632951703   
8                         91               2    142.31710767890775   
11                        92               1     77.71536034261702   
...                      ...             ...                   ...   
104992    

## Aggregate data using groupby and compute summary statistics.

In [8]:
# Data aggregation to count patient count in each state
state_pop = df.groupby('state_of_residence').size().reset_index(name='patient_count')
state_pop

state_of_residence  patient_count
0             Alabama            716
1              Alaska            696
2             Arizona            720
3            Arkansas            745
4          California            742
5            Colorado            727
6         Connecticut            756
7            Delaware            710
8             Florida            705
9             Georgia            719
10             Hawaii            667
11              Idaho            735
12           Illinois            700
13            Indiana            671
14               Iowa            773
15             Kansas            746
16           Kentucky            766
17          Louisiana            704
18              Maine            658
19           Maryland            704
20      Massachusetts            722
21           Michigan            691
22          Minnesota            774
23        Mississippi            743
24           Missouri            747
25            Montana            730
26           Nebraska            748
27             Nevada            695
28      New Hampshire            637
29         New Jersey            734
30         New Mexico            714
31           New York            748
32     North Carolina            701
33       North Dakota            726
34               Ohio            739
35           Oklahoma            683
36             Oregon            734
37       Pennsylvania            711
38       Rhode Island            727
39     South Carolina            690
40       South Dakota            682
41          Tennessee            738
42              Texas            716
43               Utah            676
44            Vermont            713
45           Virginia            693
46         Washington            698
47      West Virginia            698
48          Wisconsin            738
49            Wyoming            722

In [9]:
# Compute summary statistics
health_stats = df[['average_heart_rate','average_bp','height_(in_cm)','weight_(in_kg)']]
health_stats.describe()

average_bp
count  35828.000000
mean     109.576979
std       14.432866
min       80.000000
25%       99.000000
50%      110.000000
75%      120.000000
max      140.000000

In [10]:
# Use pivot tables or cross-tabulations for multi-dimensional analysis
pivot_table = pd.pivot_table(df, index = 'gender', columns = 'disease_diagnosed', values = ['average_bp', 'average_heart_rate'], aggfunc = 'mean')
pivot_table

<ipython-input-10-8b60b28b590c>:2: FutureWarning: pivot_table dropped a column because it failed to aggregate. This behavior is deprecated and will raise in a future version of pandas. Select only the columns that can be aggregated.
  pivot_table = pd.pivot_table(df, index = 'gender', columns = 'disease_diagnosed', values = ['average_bp', 'average_heart_rate'], aggfunc = 'mean')


average_bp                                                
disease_diagnosed     Allergy        Cold    Covid-19         Flu        None
gender                                                                       
Female             109.971489  109.861853  109.609210  109.595010  109.788110
Male               109.289529  109.777824  109.621669  109.082662  109.362260
Other              109.544885  109.544928  109.381670  109.570425  109.667936

# 3. Introduction to Alternative Libraries





In [8]:
# Loading packages
!pip install "modin[all]"
!pip install pandas==2.0
import polars as pl
import modin.pandas as mpd
import time
import pandas as pd

## Loading dataset using Polars and Modin

In [13]:
# Polars
pl_df = pl.read_csv('https://raw.githubusercontent.com/stephe-hu/datasci_2_manipulation/main/data/healthcare_data_cleaning.csv')
pl_df

shape: (105_000, 20)
┌───────────┬────────┬────────────┬────────────┬───┬────────────┬────────────┬────────────┬────────────┐
│ Patient   ┆ Gender ┆ City of    ┆ State of   ┆ … ┆ Last Visit ┆ Visit      ┆ Number of  ┆ Prescripti │
│ Age       ┆ ---    ┆ Residence  ┆ Residence  ┆   ┆ (days ago) ┆ Duration   ┆ Tests      ┆ on Cost    │
│ ---       ┆ str    ┆ ---        ┆ ---        ┆   ┆ ---        ┆ (mins)     ┆ ---        ┆ ($)        │
│ str       ┆        ┆ str        ┆ str        ┆   ┆ str        ┆ ---        ┆ str        ┆ ---        │
│           ┆        ┆            ┆            ┆   ┆            ┆ str        ┆            ┆ str        │
╞═══════════╪════════╪════════════╪════════════╪═══╪════════════╪════════════╪════════════╪════════════╡
│ 45        ┆ Other  ┆ West Brian ┆ Kentucky   ┆ … ┆ 193        ┆ 92         ┆ 5          ┆ 15.7175173 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 5532889    │
│ 25        ┆ Female ┆ East Jocel ┆ Colorado   ┆ … ┆ 195        ┆ missing    ┆ 4          ┆ 80.7864728 │
│           ┆        ┆ ynfurt     ┆            ┆   ┆            ┆            ┆            ┆ 4463952    │
│ 51        ┆ Other  ┆ South Lind ┆ Ohio       ┆ … ┆ missing    ┆ 37         ┆ 2          ┆ 64.3921393 │
│           ┆        ┆ seyland    ┆            ┆   ┆            ┆            ┆            ┆ 009105     │
│ 18        ┆ Female ┆ Taylorfort ┆ Pennsylvan ┆ … ┆ 175        ┆ 108        ┆ 2          ┆ 8.87285921 │
│           ┆        ┆            ┆ ia         ┆   ┆            ┆            ┆            ┆ 5315316    │
│ …         ┆ …      ┆ …          ┆ …          ┆ … ┆ …          ┆ …          ┆ …          ┆ …          │
│ 37        ┆ Male   ┆ Mckeetown  ┆ Utah       ┆ … ┆ 191        ┆ 47         ┆ 2          ┆ 113.867640 │
│           ┆        ┆            ┆            ┆   ┆            ┆            ┆            ┆ 25370326   │
│ 51        ┆ Female ┆ Lake Lisac ┆ Nebraska   ┆ … ┆ 172        ┆ 39         ┆ 3          ┆ 35.6224202 │
│           ┆        ┆ hester     ┆            ┆   ┆            ┆            ┆            ┆ 32475356   │
│ missing   ┆ Female ┆ Heidiborou ┆ missing    ┆ … ┆ 180        ┆ 79         ┆ missing    ┆ 124.408340 │
│           ┆        ┆ gh         ┆            ┆   ┆            ┆            ┆            ┆ 9893742    │
│ 45        ┆ Female ┆ South      ┆ Massachuse ┆ … ┆ 167        ┆ 81         ┆ missing    ┆ 87.9106014 │
│           ┆        ┆ Jeffrey    ┆ tts        ┆   ┆            ┆            ┆            ┆ 9353706    │
└───────────┴────────┴────────────┴────────────┴───┴────────────┴────────────┴────────────┴────────────┘

In [9]:
# Modin
mpd_df = pd.read_csv('https://raw.githubusercontent.com/stephe-hu/datasci_2_manipulation/main/data/healthcare_data_cleaning.csv')
mpd_df

Patient Age  Gender  City of Residence State of Residence   
0               45   Other         West Brian           Kentucky  \
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor   
0                Yes                Yes      Insurance     Dr. Williams  \
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment   
0                    Flu                 Med_E             General  \
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                    NaN                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)   
0                      66        111            183             70  \
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)   
0        65.45721578126224                   193                    92  \
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

## Comparing loading time between Modin and Polars

In [12]:
# Define the dataset path
dataset_path = 'https://raw.githubusercontent.com/stephe-hu/datasci_2_manipulation/main/data/healthcare_data_cleaning.csv'

# Using Modin
start_time = time.time()
modin_df = pd.read_csv(dataset_path)
modin_time = time.time() - start_time

# Using Polars
start_time = time.time()
polars_df = pl.read_csv(dataset_path)
polars_time = time.time() - start_time

# Print results and execution times
print(f"Modin Execution Time: {modin_time:.4f} seconds")
print(f"Polars Execution Time: {polars_time:.4f} seconds")

Modin Execution Time: 0.6878 seconds
Polars Execution Time: 0.2513 seconds


### The results show that Polars has a faster loading time (0.2513 seconds) than Modin (0.6878 seconds).